include/qpid/dispatch/internal/export.h

```diff
@@ -0,0 +1,27 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+/// @cond INTERNAL
+
+/// import/export macros
+#if defined(_WIN32)
+#  define QD_EXPORT __declspec(dllexport)
+#else
+#  define QD_EXPORT __attribute__((visibility("default"))) __attribute__((used))
```

Comment:
I was looking at this - https://developer.arm.com/documentation/dui0491/f/Compiler-specific-Features/--attribute----used---function-attribute?lang=en

It seems to say that "This function attribute informs the compiler that a static function is to be retained in the object file, even if it is unreferenced."

Is it true that __attribute__((used)) can *only* be used on *static* functions ?

I also understand from reading this - https://developer.arm.com/documentation/dui0491/f/BABCJJID - that __attribute__((used)) can also be used with  *static* variables, although I understand that this PR is not attempting to do so.

Reply:
In GCC docs it says nothing about the functions having to be static, https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes

> This attribute, attached to a function, means that code must be emitted for the function even if it appears that the function is not referenced. This is useful, for example, when the function is referenced only in inline assembly.

I do remember I had the dead code elimination issue when I tried to switch from libqpiddispatch.so to a different way of compiling the code. Because I couldn't easily have the .a library due to dead code elimination by the linker, I had to do the CMake object library, which is cumbersome in old versions of CMake.